### ISSUE TITLE
False negative with `Generic` and `Union`

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#8932]
Bound parameters are not checked in class definition.

```python
from typing import Generic

from typing_extensions import TypeVar


T = TypeVar("T", bound="AorB")


class G(Generic[T]):
    def __init__(self, obj: T) -> None:
        pass


class NotAorB:
    pass

# its, ok, validation worked as expected
Aaaa = G[NotAorB]  # Type "NotAorB" is not assignable to constrained type variable "T@G"


class A(G[NotAorB]): # No error!
    pass


class B(G[NotAorB]): # No error!
    pass


# its, ok, validation worked as expected
# Type "NotAorB" is not assignable to constrained type variable "T@G"
class C(G[NotAorB]):
    pass


AorB = A | B
```

pyright==1.1.379

**Comment by erictraut** [comment]
This is a known issue. There isn't a good solution, so I'm going to close it as "won't fix".

The problem here is that you're creating a circular reference in your code. Class `A` derives from `G` which is parameterized by `T` which has an upper bound of `A | B` which depends on `A`. Handling a circular reference like this is perhaps feasible in a traditional multi-pass type checker like mypy. Pyright uses a different architecture and evaluates all types using a JIT (just-in-time) process. This is what gives pyright good performance and low latency for language server operations. Pyright is able to partly handle this sort of a circular reference, but it needs to forego certain checks in the process, which is why you are observing no error in this case.

You can read more about this [here](https://microsoft.github.io/pyright/#/mypy-comparison?id=circular-references).

I recommend avoiding this sort of circular reference in your code.
